### ISSUE TITLE
Assign Elastic IP for EC2 Classic Instances

### ISSUE THREAD

**Issue by ssoto2** [saltstack/salt#28685]
Please Excuse me if i missed anything in the documentation. But from what i am reading only instances created within a VPC can be assigned an elastic ip address. While it is true new accounts in AWS do not offer EC2 Classic, it is still in use. 

Is there way to turn on associate elastic ip in ec2 classic deploys? If not is that planned for a feature? 

For Documentation i am using: 
salt-cloud 2015.5.3 (Lithium)

Thanks, 

S

**Comment by techhat** [comment]
Hi @ssoto2. It took me some searching, but it looks like there is a way to do it. When you set up your `network_interfaces` in your profile, you need to set `domain` to `standard` (Salt's default is `vpc`).

**Comment by rallytime** [comment]
@ssoto2 If that suggestion works for you, then we should definitely get the docs updated to include this functionality setting.

**Comment by ssoto2** [comment]
@techhat awesome! i will check that solution tonight

**Comment by ssoto2** [comment]
So that dont seem to work but i full acknowledge i might not putting this in correct. Might i ask what documentation you are looking? In Salt Clouds Docs i only see template that includes network_interface like (https://docs.saltstack.com/en/latest/topics/cloud/aws.html#cloud-profiles).  AWS API talks about Domain and setting to standard at (http://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_AllocateAddress.html) but when trying to launch an instance i get the following: 

```
salt cloud.profiles.d # salt-cloud --profile base_ec2 saltTest1109v1 -l debug        
[DEBUG   ] Reading configuration from /etc/salt/cloud
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Missing configuration file: /etc/salt/cloud.providers
[DEBUG   ] Including configuration from '/etc/salt/cloud.providers.d/aws.conf'
[DEBUG   ] Reading configuration from /etc/salt/cloud.providers.d/aws.conf
[DEBUG   ] Including configuration from '/etc/salt/cloud.providers.d/test.conf'
[DEBUG   ] Reading configuration from /etc/salt/cloud.providers.d/test.conf
[DEBUG   ] Missing configuration file: /etc/salt/cloud.profiles
[DEBUG   ] Including configuration from '/etc/salt/cloud.profiles.d/aws.conf'
[DEBUG   ] Reading configuration from /etc/salt/cloud.profiles.d/aws.conf
[DEBUG   ] Configuration file path: /etc/salt/cloud
[INFO    ] salt-cloud starting
[DEBUG   ] Could not LazyLoad parallels.avail_sizes
[DEBUG   ] LazyLoaded parallels.avail_locations
[DEBUG   ] LazyLoaded proxmox.avail_sizes
[DEBUG   ] Could not LazyLoad saltify.destroy
[DEBUG   ] Could not LazyLoad saltify.avail_sizes
[DEBUG   ] Could not LazyLoad saltify.avail_images
[DEBUG   ] Could not LazyLoad saltify.avail_locations
[DEBUG   ] LazyLoaded rackspace.reboot
[DEBUG   ] LazyLoaded openstack.list_locations
[DEBUG   ] LazyLoaded rackspace.list_locations
[DEBUG   ] Could not LazyLoad parallels.avail_sizes
[DEBUG   ] LazyLoaded parallels.avail_locations
[DEBUG   ] LazyLoaded proxmox.avail_sizes
[DEBUG   ] Could not LazyLoad saltify.destroy
[DEBUG   ] Could not LazyLoad saltify.avail_sizes
[DEBUG   ] Could not LazyLoad saltify.avail_images
[DEBUG   ] Could not LazyLoad saltify.avail_locations
[DEBUG   ] LazyLoaded rackspace.reboot
[DEBUG   ] LazyLoaded openstack.list_locations
[DEBUG   ] LazyLoaded rackspace.list_locations
[DEBUG   ] Using AWS endpoint: ec2.us-east-1.amazonaws.com
[DEBUG   ] AWS Request: https://ec2.us-east-1.amazonaws.com/?Action=DescribeInstances&Version=2014-10-01
[INFO    ] Starting new HTTPS connection (1): ec2.us-east-1.amazonaws.com
[DEBUG   ] Setting read timeout to None
[DEBUG   ] "GET /?Action=DescribeInstances&Version=2014-10-01 HTTP/1.1" 200 None
[DEBUG   ] AWS Response Status Code: 200
[DEBUG   ] Generating minion keys for 'saltTest1109v1'
[DEBUG   ] MasterEvent PUB socket URI: ipc:///var/run/salt/master/master_event_pub.ipc
[DEBUG   ] MasterEvent PULL socket URI: ipc:///var/run/salt/master/master_event_pull.ipc
[DEBUG   ] Sending event - data = {'_stamp': '2015-11-09T21:31:44.344054'}
[DEBUG   ] Sending event - data = {'profile': 'base_ec2', 'event': 'starting create', '_stamp': '2015-11-09T21:31:44.345146', 'name': 'saltTest1109v1', 'provider': 'aws-us-east:ec2'}
[INFO    ] Creating Cloud VM saltTest1109v1 in us-east-1
[DEBUG   ] Create network interface: {'DeviceIndex': 0, 'domain': 'standard', 'AssociatePublicIpAddress': True}
[DEBUG   ] Using AWS endpoint: ec2.us-east-1.amazonaws.com
[DEBUG   ] AWS Request: https://ec2.us-east-1.amazonaws.com/?Action=DescribeSubnets&Version=2014-10-01
[INFO    ] Starting new HTTPS connection (1): ec2.us-east-1.amazonaws.com
[DEBUG   ] Setting read timeout to None
[DEBUG   ] "GET /?Action=DescribeSubnets&Version=2014-10-01 HTTP/1.1" 200 None
[DEBUG   ] AWS Response Status Code: 200
[ERROR   ] Failed to create VM saltTest1109v1. Configuration value 'SubnetId' needs to be set
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/cloud/__init__.py", line 1241, in create
    output = self.clouds[func](vm_)
  File "/usr/lib/python2.7/dist-packages/salt/cloud/clouds/ec2.py", line 2153, in create
    data, vm_ = request_instance(vm_, location)
  File "/usr/lib/python2.7/dist-packages/salt/cloud/clouds/ec2.py", line 1544, in request_instance
    _new_eni = _create_eni(interface)
  File "/usr/lib/python2.7/dist-packages/salt/cloud/clouds/ec2.py", line 1152, in _create_eni
    if subnet['subnetId'] == interface['SubnetId']:
KeyError: 'SubnetId'
Error: There was a profile error: Failed to deploy VM
```

My Profile: 

```
base_ec2:
  provider: aws-us-east
  image: ami-8a720fe0
  size: r3.large
  ssh_username: ec2-user
  network_interfaces:
    - DeviceIndex: 0
      domain: stanard
      #auto assign public ip (not EIP)
      AssociatePublicIpAddress: True
```

**Comment by techhat** [comment]
Is that `stanard` typo in the issue also in your local profile?

**Comment by ssoto2** [comment]
no  i wish that was the case.. Just my typo

as per 

[DEBUG   ] Create network interface: {'DeviceIndex': 0, 'domain': 'standard', 'AssociatePublicIpAddress': True}

**Comment by rallytime** [comment]
@ssoto2 I am not very familiar with this code, but I am wondering if https://github.com/saltstack/salt/pull/20263 might be helpful here. This particular change is in the 2015.8.x releases, but not the 2015.5.x releases. Are you in a position to apply that change and give it a shot? Or perhaps upgrade to 2015.8.3 to see if `domain: standard` works there with the way you've used `network_interfaces`?

The only other thing I can think of to get past that stack trace is to define a `SubnetId ` in your `network_interfaces` config. However, given that you're setting `AssociatePublicIpAddress: True`, I think the first option I mentioned is likely a better way to approach this. Unless @techhat has some other input.

**Comment by rallytime** [comment]
@ssoto2 Actually, I'm sorry, I think https://github.com/saltstack/salt/pull/20972 is actually the fix you want, which should be available in the latest 2015.5.x release as well as 2015.8.3.

**Comment by rallytime** [comment]
Since we haven't heard back here, I am going to close this. If this remains an issue after considering the above comments, please let us know and we'll be glad to take another look!
